---
title: "WGCNA_book chapter"
author: "Dae Kwan Ko"
date: "11/1/2021"
output: html_document
---

## R Markdown

```{r setup, include=FALSE, echo=FALSE}
require("knitr")
opts_knit$set(root.dir = "/Users/LOCAL DIRECTORY")
# By default, the working directory for R code chunks is the directory that contains the Rmd document. 
rm(list=ls()) # Remove everything from the R environment. Important if you have some variables saved already
# Display the current working directory
getwd();
```


```{r loadLibs}
# if (!requireNamespace("BiocManager", quietly = TRUE))
# install.packages("BiocManager")
# BiocManager::install("impute")
# BiocManager::install("preprocessCore")
# BiocManager::install("GO.db")
# BiocManager::install("AnnotationDbi")
# BiocManager::install("WGCNA")
require(WGCNA)
require(flashClust)
require(RColorBrewer)
require(RColorBrewer)
require(gplots)
require(ggplot2)
require(pheatmap)
## Load the required libraries.
# lnames = load(file = "/Users/dkko/Google Drive/My Drive/Research/1_Brandizzi Lab_MSU/*Project_GLBRC/1_Gene regulation dynamics in sorghum stem/RNAseq_sorghum.stem_TPJ2016/WGCNA/WGCNA_full.data.RData");
#The variable lnames contains the names of loaded variables.
# lnames
```

# http://pklab.med.harvard.edu/scw2014/WGCNA.html
# Tutorials: https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/, http://pklab.med.harvard.edu/scw2014/WGCNA.html 
# FPKM or log2(1+FPKM): https://www.biostars.org/p/280650/ 

```{r input.file.prep}
options(stringsAsFactors = FALSE)
df <- read.table("wgcna_input_log2fc.txt", header=TRUE, sep ="\t")
dim(df)
names(df)
rnames <- df[,1]
rnames
rownames(df) <- rnames
FPKM_DEGs <- df
names(FPKM_DEGs)
datExpr = as.data.frame(t(FPKM_DEGs[, -c(1)]))
View(datExpr)
dim(datExpr)
```

# Checking data for excessive missing values and identification of outlier microarray samples
https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/Tutorials/FemaleLiver-01-dataInput.pdf
```{r input.file.prep}
# We first check for genes and samples with too many missing values:
gsg = goodSamplesGenes(datExpr, verbose = 3);
gsg$allOK
# If the last statement returns TRUE, all genes have passed the cuts. If not, we remove the offending genes and samples from the data:
if (!gsg$allOK)
{
  # Optionally, print the gene and sample names that were removed:
  if (sum(!gsg$goodGenes)>0)
     printFlush(paste("Removing genes:", paste(names(datExpr)[!gsg$goodGenes], collapse = ", ")));
  if (sum(!gsg$goodSamples)>0)
     printFlush(paste("Removing samples:", paste(rownames(datExpr)[!gsg$goodSamples], collapse = ", ")));
  # Remove the offending genes and samples from the data:
  datExpr = datExpr[gsg$goodSamples, gsg$goodGenes]
}
```

Gene coexpression networks are constructed using the WGCNA package run using batch mode to get a set of eigengenes.

This first part can be run if desired. It is used to determine a good power value to use in network construction.
# Choosing the soft-thresholding power: analysis of network topology
```{r WGCNA_Power}
# Choose a set of soft-thresholding power
powers = c(c(1:10), seq(from = 12, to=20, by=2))
# Call the network topology analysis function
# sft=pickSoftThreshold(datExpr,dataIsExpr=TRUE, powerVector=powers, corFnc=cor, corOptions=list(use='p', method='spearman'), networkType="signed", verbose = 5)
# above script made computer frozen
sft=pickSoftThreshold(datExpr, powerVector=powers, networkType="signed", verbose = 5)
## link for unsigned or signed http://www.peterlangfelder.com/signed-or-unsigned-which-network-type-is-preferable/

# Plot the results:
sizeGrWindow(9, 5)
par(mfrow = c(1,2))
cex1 = 0.9

# Scale-free topology fit index as a function of the soft-thresholding power
plot(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],xlab="Soft Threshold (power)",ylab="Scale Free Topology Model Fit, signed R^2",type="n", main = paste("Scale independence"))
text(sft$fitIndices[,1], -sign(sft$fitIndices[,3])*sft$fitIndices[,2],labels=powers,cex=cex1,col="red")
# this line corresponds to using an R^2 cut-off of h
abline(h=0.60,col="red")

# Mean connectivity as a function of the soft-thresholding power
plot(sft$fitIndices[,1], sft$fitIndices[,5],xlab="Soft Threshold (power)",ylab="Mean Connectivity", type="n",main = paste("Mean connectivity"))
text(sft$fitIndices[,1], sft$fitIndices[,5], labels=powers, cex=cex1,col="red")
```

what power needs to be selected? 
# https://www.biostars.org/p/359179/ # generally, the first past 0.9
# https://www.biostars.org/p/319355/
# https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/faq.html 

I have decided 18
Now construct the networks. 

# Co-expression similarity and adjacency
```{r WGCNA_Power}
# We now calculate the adjacencies, using the soft thresholding power 16:
## generating adjacency and TOM similarity matrices based on the selected softpower
softPower = 18
## calcluate the adjacency matrix
adjacency= adjacency(datExpr,type = "signed", power = softPower)
# Turn adjacency into topological overlap
TOM = TOMsimilarity(adjacency);
dissTOM = 1-TOM

# Clustering using TOM
## We now use hierarchical clustering to produce a hierarchical clustering tree (dendrogram) of genes. Note that we use the function hclust that provides a much faster hierarchical clustering routine than the standard hclust function.
# Call the hierarchical clustering function
geneTree = flashClust(as.dist(dissTOM),method="average")

# plot the resulting clustering tree (dendrogram)
sizeGrWindow(12,9)
plot(geneTree, xlab="", sub="", main = "Gene clustering on TOM-based dissimilarity",
    labels = FALSE, hang = 0.04);

# Set the minimum module size
minModuleSize = 25;

# Module identification using dynamic tree cut
dynamicMods = cutreeDynamic(dendro = geneTree,  method="tree", minClusterSize = minModuleSize);

# the following command gives the module labels and the size of each module. Lable 0 is reserved for unassigned genes. largest to smallest
table(dynamicMods) # Label 0 is reserved for unassigned genes.
```
```{r plot}
# We now plot the module assignment under the gene dendrogram:
# Plot the module assignment under the dendrogram; note: The grey color is reserved for unassigned genes
# Convert numeric lables into colors
dynamicColors = labels2colors(dynamicMods)
table(dynamicColors)
# Plot the dendrogram and colors underneath
sizeGrWindow(8,6)
plotDendroAndColors(geneTree, dynamicColors, "Dynamic Tree Cut", dendroLabels = FALSE, hang = 0.03, addGuide = TRUE, guideHang = 0.05, main = "Gene dendrogram and module colors")
```

# Merging of modules whose expression profiles are very similar
The Dynamic Tree Cut may identify modules whose expression profiles are very similar. It may be prudent to merge such modules since their genes are highly co-expressed. To quantify co-expression similarity of entire modules, we calculate their eigengenes and cluster them on their correlation:
```{r merging}
# Calculate eigengenes
MEList = moduleEigengenes(datExpr, colors = dynamicColors)
MEs = MEList$eigengenes
# Calculate dissimilarity of module eigengenes
MEDiss = 1-cor(MEs);
# Cluster module eigengenes
METree = hclust(as.dist(MEDiss), method = "average");
# Plot the result
sizeGrWindow(7, 6)
plot(METree, main = "Clustering of module eigengenes",
     xlab = "", sub = "")


MEDissThres = 0.05
# Plot the cut line into the dendrogram
abline(h=MEDissThres, col = "red")
# Call an automatic merging function
merge = mergeCloseModules(datExpr, dynamicColors, cutHeight = MEDissThres, verbose = 3)
# The merged module colors
mergedColors = merge$colors;
# Eigengenes of the new merged modules:
mergedMEs = merge$newMEs;


# To see what the merging did to our module colors, we plot the gene dendrogram again, with the original and merged module colors underneath
sizeGrWindow(12, 9)
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),
                   c("Dynamic Tree Cut", "Merged dynamic"),
                   dendroLabels = FALSE, hang = 0.03,
                   addGuide = TRUE, guideHang = 0.05)

pdf(file = "dendrogram.pdf", width = 7, height = 6) # defaults to 7 x 7 inches
plotDendroAndColors(geneTree, cbind(dynamicColors, mergedColors),
                   c("Dynamic Tree Cut", "Merged dynamic"),
                   dendroLabels = FALSE, hang = 0.03,
                   addGuide = TRUE, guideHang = 0.05)
dev.off()

# In the subsequent analysis, we will use the merged module colors in mergedColors. We save the relevant variables for use in subsequent parts of the tutorial:
# Rename to moduleColors
moduleColors = mergedColors
```


# Visualizing the gene network
# this takes LOTS OF TIME
```{r visuzalizing}
# calculate topological overlap anew: this could be done more efficiently by saving the TOM
# calculated during module detection, but let us do it again here.
dissTOM = 1-TOMsimilarityFromExpr(datExpr, power = 18);
# Transform dissTOM with a power to make moderately strong connections more visible in the heatmap
plotTOM = dissTOM^7;
# Set diagonal to NA for a nicer plot
diag(plotTOM) = NA;
# Call the plot function
sizeGrWindow(9,9)
TOMplot(plotTOM, geneTree, moduleColors, main = "Network heatmap plot, all genes")

# export
# to save tiff file, since pdf version does not look fine
# tiff(file = "/Users/dkko/Google Drive/My Drive/Research/1_Brandizzi Lab_MSU/Other manuscripts/Book chapter method_Plant Proteostasis/WGCNA practice/TOMplot_v2.tiff", width = 2000, height = 2000, units = "px", res = 400)
# TOMplot(plotTOM, geneTree, moduleColors, main = "Network heatmap plot, all genes")
# dev.off()
```


# Visualizing the network of eigengenes
```{r visuzalizing}
# Recalculate module eigengenes
MEs = moduleEigengenes(datExpr, moduleColors)$eigengenes
plotEigengeneNetworks(MEs, "", marDendro = c(0,4,1,2), marHeatmap = c(3,4,1,2))
# The function produces a dendrogram of the eigengenes and trait(s), and a heatmap of their relationships.
pdf(file = "plotEigengeneNetworks.pdf", width = 7, height = 6) # defaults to 7 x 7 inches
plotEigengeneNetworks(MEs, "", marDendro = c(0,4,1,2), marHeatmap = c(3,4,1,2))
dev.off()

# To split the dendrogram and heatmap plots, we can use the following code
# Plot the dendrogram
sizeGrWindow(6,6);
par(cex = 1.0)
plotEigengeneNetworks(MEs, "Eigengene dendrogram", marDendro = c(0,4,2,0),
                    plotHeatmaps = FALSE)
# Plot the heatmap matrix (note: this plot will overwrite the dendrogram plot)
par(cex = 1.0)
plotEigengeneNetworks(MEs, "Eigengene adjacency heatmap", marHeatmap = c(3,4,2,2),
                    plotDendrograms = FALSE, xLabelsAngle = 90)

# to see # of genes in each module
table(mergedColors)

# export eigengenes in a table
write.table(MEs, file = "/Users/dkko/Google Drive/My Drive/Research/1_Brandizzi Lab_MSU/*Project_GLBRC/1_Gene regulation dynamics in sorghum stem/RNAseq_sorghum.stem_TPJ2016/WGCNA/Sorghum_WGCNA_eigengenes.txt", sep = "\t")
```
# Output gene lists
```{r exporting.list}
# Recalculate module eigengenes
MEs = moduleEigengenes(datExpr, moduleColors)$eigengenes


for (color in moduleColors){
  module=datExpr[which(moduleColors==color)]
  write.table(module, paste("module_",color, ".txt",sep=""), sep="\t", row.names=TRUE, col.names=TRUE,quote=FALSE)
}
```

# Exporting network data to network visualization software
# Exporting to Cytoscape
# Cytoscape allows the user to input an edge file and a node file, allowing the user to specify for example the link weights and the node colors. Here we demonstrate the output of two modules, the red and brown ones, to Cytoscape.
```{r exporting.to.cytoscape}
# Recalculate topological overlap if needed
#TOM = TOMsimilarityFromExpr(datExpr, power = 12);
# modules
table(mergedColors)
# Select modules
modules = c("lightcyan");
# Select module probes
probes = names(datExpr)
inModule = is.finite(match(moduleColors, modules));
modProbes = probes[inModule];
#modGenes = annot$gene_symbol[match(modProbes, annot$substanceBXH)];
# Select the corresponding Topological Overlap
modTOM = TOM[inModule, inModule];
dimnames(modTOM) = list(modProbes, modProbes)

cyt = exportNetworkToCytoscape(modTOM,
                               edgeFile = paste("CytoscapeInput-edges-", paste(modules, collapse="-"), ".txt", sep=""),
                               nodeFile = paste("CytoscapeInput-nodes-", paste(modules, collapse="-"), ".txt", sep=""),
                               weighted = TRUE,
                               threshold = 0.2,
                               nodeNames = modProbes,
                               nodeAttr = moduleColors[inModule]);
# https://www.biostars.org/p/286774/
  
  
  
#####some useful links regarding to eigengene expression
# https://www.biostars.org/p/113125/
```
